sources/engine/Stride.VirtualReality/OculusOVR/OculusOverlay.cs

```diff
@@ -36,7 +39,11 @@ public OculusOverlay(IntPtr ovrSession, GraphicsDevice device, int width, int he
                 }
 
                 textures[i] = new Texture(device);
-                textures[i].InitializeFromImpl(new Texture2D(ptr), false);
+                unsafe
+                {
+                    var tex = new ID3D11Texture2D((void**)ptr);
+                    textures[i].InitializeFromImpl(new ComPtr<ID3D11Texture2D>(&tex), false);
```

Comment:
A general note about `ComPtr`s (not a problem here, but wanted to ensure I put this somewhere): the idea of `ComPtr` is that it does some of the management of the underlying pointer's lifetime for you.

So:
```cs
ID3D11Texture2D* nativePtr = /* some code to get a pointer from native */;
// let's call the refcount of nativePtr here "n"
var ptr1 = new ComPtr<ID3D11Texture2D>(nativePtr); // calls AddRef, meaning the refcount is now n + 1
var ptr2 = new ComPtr<ID3D11Texture2D>(ptr1); // copies ptr1 and calls AddRef again, meaning the refcount is now n + 2
ptr1.Dispose(); // Release is called, decrementing the refcount to n + 1
ptr2.Dispose(); // Release is called, decrementing the refcount back to n
// if n is zero, the native pointer is freed here.
```

It's best to model this in your code as closely as possible. C# unfortunately does not have copy constructors like C++ does, otherwise this would be a lot easier.

As such, whenever you're passing a `ComPtr` to a method or something, it's best to clone that `ComPtr` as illustrated above (`ptr2` being created from `ptr1`). Likewise, to avoid memory leaks always ensure you dispose your `ComPtr`s.

Reply:
I understand a bit better, i think i did some mistakes in other parts i will recheck everything to be sure.
If i understood well, everytime i need to "borrow" a `ComPtr` i should make sure to use it and dispose it after ward ?

Reply:
Yeah. How the C++ implementation does it is every time the struct is copied, the ref counter goes up and every time the struct goes out of scope, the ref counter goes down.

```cpp
ComPtr<ID3D11Texture2D> thing_one { native_ptr };
{
    ComPtr<ID3D11Texture2D> thing_two { thing_one }; // C++ magic increments the ref counter here because it's a copy
} // thing_two is out of scope now, so ref counter decrements
```

Obviously it's not as easy to do that in C# so the closest you can get is:
- where you would otherwise copy: `new(ptr)` to create a copy of the ptr.
- at the end of methods which take a `ComPtr`: `ptr.Dispose()`

Perhaps that's not the cleanest way, and feel free to do what fits best; but hopefully my explanation on how the C++ original works helps :)